processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java

```diff
@@ -0,0 +1,249 @@
+      }
+    };
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return (((SerializablePair<Long, Double>) lhs).lhs <= ((SerializablePair<Long, Double>) rhs).lhs) ? lhs : rhs;
+  }
+
+  @Override
+  public AggregatorFactory getCombiningFactory()
+  {
+    return new DoubleFirstAggregatorFactory(name, name)
+    {
+      @Override
+      public Aggregator factorize(ColumnSelectorFactory metricFactory)
+      {
+        final ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(name);
+        return new DoubleFirstAggregator(name, null, null)
+        {
+          @Override
+          public void aggregate()
+          {
+            SerializablePair<Long, Double> pair = (SerializablePair<Long, Double>) selector.get();
+            if (pair.lhs < firstTime) {
+              firstTime = pair.lhs;
+              firstValue = pair.rhs;
+            }
+          }
+        };
+      }
+
+      @Override
+      public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+      {
+        final ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(name);
+        return new DoubleFirstBufferAggregator(null, null)
+        {
+          @Override
+          public void aggregate(ByteBuffer buf, int position)
+          {
+            SerializablePair<Long, Double> pair = (SerializablePair<Long, Double>) selector.get();
+            long firstTime = buf.getLong(position);
+            if (pair.lhs < firstTime) {
+              buf.putLong(position, pair.lhs);
+              buf.putDouble(position + Longs.BYTES, pair.rhs);
+            }
+          }
+        };
+      }
+    };
+  }
+
+  @Override
+  public AggregatorFactory getMergingFactory(AggregatorFactory other) throws AggregatorFactoryNotMergeableException
+  {
+    if (other.getName().equals(this.getName()) && this.getClass() == other.getClass()) {
+      return getCombiningFactory();
+    } else {
+      throw new AggregatorFactoryNotMergeableException(this, other);
+    }
+  }
+
+  @Override
+  public List<AggregatorFactory> getRequiredColumns()
+  {
+    return Arrays.<AggregatorFactory>asList(new DoubleFirstAggregatorFactory(fieldName, fieldName));
+  }
+
+  @Override
+  public Object deserialize(Object object)
+  {
+    Map map = (Map) object;
+    return new SerializablePair<>(((Number) map.get("lhs")).longValue(), ((Number) map.get("rhs")).doubleValue());
+  }
+
+  @Override
+  public Object finalizeComputation(Object object)
+  {
+    return ((SerializablePair<Long, Double>) object).rhs;
+  }
+
+  @Override
+  @JsonProperty
+  public String getName()
+  {
+    return name;
+  }
+
+  @JsonProperty
+  public String getFieldName()
+  {
+    return fieldName;
+  }
+
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
```

Comment:
add a `0xFF` after this byte array when putting it in the byte buffer to ensure proper string termination.